sql: short-living non-stopped monitors on ungraceful shutdown

pkg/sql/logictest/tests/fakedist/fakedist_test.TestLogic_tenant [failed](https://tanzanite.cluster.engflow.com/invocations/default/40f805a9-cdee-42b9-9fd5-5d565b3f60f3?testReportRun=18&testReportShard=45&testReportAttempt=1#targets-Ly9wa2cvc3FsL2xvZ2ljdGVzdC90ZXN0cy9mYWtlZGlzdDpmYWtlZGlzdF90ZXN0) on master @ [99c0d277c26344a08903e1372ab62a034336d178](https://github.com/cockroachdb/cockroach/commits/99c0d277c26344a08903e1372ab62a034336d178):

Fatal error:

```
panic: found 1 short-living non-stopped monitors in in-mem temp storage
in-mem temp storage 0 B (512 MiB / 0 B)
    flow-disk-monitor 0 B
 [recovered]
	panic: found 1 short-living non-stopped monitors in in-mem temp storage
in-mem temp storage 0 B (512 MiB / 0 B)
    flow-disk-monitor 0 B
 [recovered]
	panic: found 1 short-living non-stopped monitors in in-mem temp storage
in-mem temp storage 0 B (512 MiB / 0 B)
    flow-disk-monitor 0 B


```
Stack: 

```
goroutine 94566 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x475de5?, {0x71799b8, 0xc0185ee720})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x65
panic({0x550c2e0?, 0xc00c1e6a20?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x10db58a?, {0x71799b8, 0xc0185ee720})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x65
panic({0x550c2e0?, 0xc00c1e6a20?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc015ddf0e0, {0x71799b8, 0xc0070794a0}, 0x0)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657 +0x653
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).EmergencyStop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:593
github.com/cockroachdb/cockroach/pkg/server.newSQLServer.func6()
	github.com/cockroachdb/cockroach/pkg/server/server_sql.go:862 +0x2c
github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0xc016207ac0?)
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:108 +0x12
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc00ad65a70, {0x71799b8, 0xc0185ee720})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:550 +0x131
github.com/cockroachdb/cockroach/pkg/server.(*channelOrchestrator).propagateUngracefulStopAsync.func1({0x71799b8, 0xc0185ee720})
	github.com/cockroachdb/cockroach/pkg/server/server_controller_channel_orchestrator.go:489 +0x15c
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:485 +0x13a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 94565
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:476 +0x3fe
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestLogic_tenant
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestLogic_tenant2576607349
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:144: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
[05:51:57] setting distsql_workmem='54447B';
=== RUN   TestLogic_tenant/drop_tenant
=== RUN   TestLogic_tenant/read_only
=== RUN   TestLogic_tenant/reclaim_name
=== RUN   TestLogic_tenant/regression_105115
=== RUN   TestLogic_tenant/restrict_system_access
[05:51:59] --- done: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/fakedist/fakedist_test_/fakedist_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/tenant with config fakedist: 117 tests, 0 failures
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=18</code>
 - <code>shard=45</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_tenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39107

--- a/pkg/ccl/changefeedccl/changefeed_processors.go
+++ b/pkg/ccl/changefeedccl/changefeed_processors.go
@@ -399,7 +399,10 @@ func (ca *changeAggregator) startKVFeed(
 	opts changefeedbase.StatementOptions,
 ) (kvevent.Reader, chan struct{}, chan error, error) {
 	cfg := ca.flowCtx.Cfg
-	kvFeedMemMon := mon.NewMonitorInheritWithLimit("kvFeed", memLimit, parentMemMon, false /* longLiving */)
+	// CDC DistSQL flows are long-living, so we mark the memory monitors
+	// accordingly.
+	const longLiving = true
+	kvFeedMemMon := mon.NewMonitorInheritWithLimit("kvFeed", memLimit, parentMemMon, longLiving)
 	kvFeedMemMon.StartNoReserved(ctx, parentMemMon)
 	buf := kvevent.NewThrottlingBuffer(
 		kvevent.NewMemBuffer(kvFeedMemMon.MakeBoundAccount(), &cfg.Settings.SV, &ca.metrics.KVFeedMetrics),

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -718,7 +718,7 @@ func NewServer(cfg Config, stopper *stop.Stopper) (serverctl.ServerStartupInterf
 		//
 		// Note that we don't do this for SQL servers of tenants since there we
 		// can have ungraceful shutdown whenever the node is quiescing, so we
-		// have some short-living monitors that are't stopped.
+		// have some short-living monitors that aren't stopped.
 		sqlMonitorAndMetrics.rootSQLMemoryMonitor.EmergencyStop(ctx)
 	}))
 

--- a/pkg/server/server_sql.go
+++ b/pkg/server/server_sql.go
@@ -858,9 +858,17 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 		RootSQLMemoryPoolSize:    cfg.MemoryPoolSize,
 	}
 	cfg.TempStorageConfig.Mon.SetMetrics(distSQLMetrics.CurDiskBytesCount, distSQLMetrics.MaxDiskBytesHist)
-	cfg.stopper.AddCloser(stop.CloserFn(func() {
-		cfg.TempStorageConfig.Mon.EmergencyStop(ctx)
-	}))
+	if codec.ForSystemTenant() {
+		// Stop the temp storage disk monitor to enforce (in test builds) that
+		// all short-living descendants are stopped too.
+		//
+		// Note that we don't do this for SQL servers of tenants since there we
+		// can have ungraceful shutdown whenever the node is quiescing, so we
+		// have some short-living monitors that aren't stopped.
+		cfg.stopper.AddCloser(stop.CloserFn(func() {
+			cfg.TempStorageConfig.Mon.EmergencyStop(ctx)
+		}))
+	}
 	if distSQLTestingKnobs := cfg.TestingKnobs.DistSQL; distSQLTestingKnobs != nil {
 		distSQLCfg.TestingKnobs = *distSQLTestingKnobs.(*execinfra.TestingKnobs)
 	}

--- a/pkg/util/mon/bytes_usage.go
+++ b/pkg/util/mon/bytes_usage.go
@@ -653,13 +653,12 @@ func (mm *BytesMonitor) doStop(ctx context.Context, check bool) {
 				// Ignore mm itself if it is short-living.
 				numShortLiving--
 			}
-			// TODO(#124848): uncomment this.
-			//if numShortLiving > 0 {
-			//	panic(errors.AssertionFailedf(
-			//		"found %d short-living non-stopped monitors in %s\n%s",
-			//		numShortLiving, mm.name, sb.String(),
-			//	))
-			//}
+			if numShortLiving > 0 {
+				panic(errors.AssertionFailedf(
+					"found %d short-living non-stopped monitors in %s\n%s",
+					numShortLiving, mm.name, sb.String(),
+				))
+			}
 		}
 	}
 

